src/StdUtils.sol

```diff
@@ -100,11 +114,32 @@ abstract contract StdUtils {
         return addressFromLast20Bytes(keccak256(abi.encodePacked(bytes1(0xff), deployer, salt, initcodeHash)));
     }
 
-    function bytesToUint(bytes memory b) internal pure virtual returns (uint256) {
-        require(b.length <= 32, "StdUtils bytesToUint(bytes): Bytes length exceeds 32.");
-        return abi.decode(abi.encodePacked(new bytes(32 - b.length), b), (uint256));
+    // Performs a single call with Multicall3 to query the ERC-20 token balances of the given addresses.
+    function getTokenBalances(address token, address[] memory addresses) internal returns (uint256[] memory balances) {
```

Comment:
Tests for this method would be great 🙏

Reply:
I will write a few later this week.

Reply:
@mds1 I started testing this function but there must be something that I misunderstand about Multicall or fork testing or both because I cannot get the revert test to pass ..

I wrote this:

```solidity
modifier forkMainnet() {
    vm.createSelectFork({urlOrAlias: "mainnet", blockNumber: 16376000});
    _;
}

function testCannotGetTokenBalances_NonTokenContract() external forkMainnet {
    address token = address(1729);
    address[] memory addresses = new address[](1);
    addresses[0] = USDC_HOLDER;
    vm.expectRevert("Multicall3: call failed");
    getTokenBalances(token, addresses);
}
```

The test fails though I think it should pass because there is no token contract at address 1729 on Ethereum Mainnet .. the call [should revert with "Multicall3: call failed"](https://github.com/mds1/multicall/blob/eb34ad2954f9ceb475a24bb0155bff3bef0f5409/src/Multicall3.sol#L50), but it doesn't.

I also tried replacing the expect revert with the following:

```solidity
vm.expectRevert();
vm.expectRevert(bytes(""));
```

Nothing worked.

Reply:
This is working as expected actually. A tx that sends calldata to an address without code is considered a successful transaction by the EVM and it doesn't revert. Normally if doing this in solidity, the revert comes from solidity: the `IERC20` interface (for example) that you defined says `balanceOf` should return a uint256, and when solidity sees that no data was returned because you called an address with no code, it reverts. 

You can probably solve this by instead calling a solidity contract that has code, but doesn't implement the `balanceOf` function and has no fallback method (maybe the uniswap factory would work for this?). The calldata will not match anything in the function dispatcher and the called contract should revert as a result

Reply:
> A tx that sends calldata to an address without code is considered a successful transaction by the EVM and it doesn't revert.

TIL 😬

> You can probably solve this by instead calling a solidity contract that has code, but doesn't implement the balanceOf function and has no fallback method

Great idea, will do 🫡

Reply:
Any update here? Planning to do a new forge-std release once this PR gets in, would love to do it this week if possible

Reply:
Sorry @mds1, I started working on this last week but I bumped into a blocker. I should have posted an update here.

I picked up my work again, and I still bumping into the same blocker.

The issue is related to forking mainnet. To test `getTokenBalance`, I need to rewrite the `StdUtils` tests to contain a modifier for forking mainnet, and I also need to rewrite the test contract to call a mock rather than an internal function (If I don't this, the function does not revert on `multicall.aggregate(calls)` - it reverts due to an "index out of bounds" error trying to access the returned data which is nil).

I wrote this:

```solidity
    StdUtilsMock internal utils = new StdUtilsMock();

    address internal USDC_HOLDER = 0xDa9CE944a37d218c3302F6B82a094844C6ECEb17;

    modifier forkMainnet() {
        vm.createSelectFork({urlOrAlias: "mainnet", blockNumber: 16376000});
        _;
    }

    function testCannotGetTokenBalances_NonTokenContract() external forkMainnet {
        // The UniswapV2Factory contract has neither a "balanceOf" function nor a fallback function.
        address token = address(0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f);
        address[] memory addresses = new address[](1);
        addresses[0] = USDC_HOLDER;
        vm.expectRevert();
        utils.getTokenBalances_(token, addresses);
    }
```

But I cannot for the life of me understand why it doesn't work. Even if you add a dummy `foo` function in the `StdUtilsMock` contract, and call it in the test function, the call still reverts.

Might the problem be related to the deployment of the `utils` mock? Can deployments not be made when forking?

Or is the issue related to the RPC URL? I tried setting my own key, but I observed the same behavior.

Reply:
Just undid your last commit, then pushed a commit which seems to work as expected!

Reply:
Thank you!

The issue was that I was deploying the mock contract at the contract level (right underneath `StdUtilsTest`), which means the mock was being deployed to the local network.

You added it in the test function body, where it gets deployed to the mainnet fork - as it should.

Reply:
@mds1 - I added more tests:

- EOA as an input, should revert
- USDC holders
- SHIB holders

The PR should be ready for merging, pending your review of course.

Reply:
Pushed two small commits to fix a typo + resolve failing CI, should be good to go once it passes. Thanks a lot for this! 🙌

Reply:
Awesome stuff, thanks @mds1 for the quick review. And yeah sorry, I keep forgetting about supporting older solc versions in my PRs.